src/orion/algo/evolution_es.py

```diff
@@ -0,0 +1,263 @@
+
+        if evolution_es.mutate_attr:
+            self.mutate_attr = copy.deepcopy(evolution_es.mutate_attr)
+        else:
+            self.mutate_attr = {}
+
+        function_string = self.mutate_attr.pop('function',
+                                               "orion.algo.mutate_functions.default_mutate")
+        mod_name, func_name = function_string.rsplit('.', 1)
+        mod = importlib.import_module(mod_name)
+        self.mutate_func = getattr(mod, func_name)
+
+        for i in range(len(space.values())):
+            if not i == self.eves.fidelity_index:
+                self.search_space_remove_fidelity.append(i)
+
+    def get_candidates(self, rung_id):
+        """Get a candidate for promotion"""
+        if self.has_rung_filled(rung_id + 1):
+            return []
+
+        rung = self.rungs[rung_id]['results']
+
+        population_range = (self.eves.nums_population
+                            if len(list(rung.values())) > self.eves.nums_population
+                            else len(list(rung.values())))
+
+        for i in range(population_range):
+            for j in self.search_space_remove_fidelity:
+                self.eves.population[j][i] = list(rung.values())[i][1][j]
+            self.eves.performance[i] = list(rung.values())[i][0]
+
+        population_index = list(range(self.eves.nums_population))
+        red_team = np.random.choice(population_index, self.eves.nums_comp_pairs, replace=False)
+        diff_list = list(set(population_index).difference(set(red_team)))
+        blue_team = np.random.choice(diff_list, self.eves.nums_comp_pairs, replace=False)
+
+        winner_list = []
+        loser_list = []
+
+        hurdles = 0
+        for i, _ in enumerate(red_team):
+            winner = (red_team
+                      if self.eves.performance[red_team[i]] < self.eves.performance[blue_team[i]]
+                      else blue_team)
+            loser = (red_team
+                     if self.eves.performance[red_team[i]] >= self.eves.performance[blue_team[i]]
+                     else blue_team)
+
+            winner_list.append(winner[i])
+            loser_list.append(loser[i])
+            hurdles += self.eves.performance[winner[i]]
+            self._mutate(winner[i], loser[i])
+
+        hurdles /= len(red_team)
+        self.eves.hurdles.append(hurdles)
+
+        logger.debug('Evolution hurdles are: %s', str(self.eves.hurdles))
+
+        points = []
+        for i in range(population_range):
+            point = [0] * len(self.space)
+            nums_all_equal = 0
+            while True:
+                point[self.eves.fidelity_index] = \
+                    list(rung.values())[i][1][self.eves.fidelity_index]
+
+                for j in self.search_space_remove_fidelity:
+                    if self.space.values()[j].type == "integer" or \
+                       self.space.values()[j].type == "categorical":
+                        point[j] = int(self.eves.population[j][i])
```

Comment:
Why is this casting necessary, shouldn't it be part of the mutate function?

Reply:
Fixed.